src/main/java/eu/dissco/core/digitalspecimenprocessor/service/EqualityService.java

```diff
@@ -0,0 +1,177 @@
+
+import static com.jayway.jsonpath.Criteria.where;
+import static com.jayway.jsonpath.Filter.filter;
+import static com.jayway.jsonpath.JsonPath.using;
+import static eu.dissco.core.digitalspecimenprocessor.domain.EntityRelationshipType.HAS_MEDIA;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.github.fge.jsonpatch.diff.JsonDiff;
+import com.jayway.jsonpath.Configuration;
+import com.jayway.jsonpath.DocumentContext;
+import eu.dissco.core.digitalspecimenprocessor.domain.media.DigitalMediaProcessResult;
+import eu.dissco.core.digitalspecimenprocessor.domain.specimen.DigitalSpecimenEvent;
+import eu.dissco.core.digitalspecimenprocessor.domain.specimen.DigitalSpecimenWrapper;
+import eu.dissco.core.digitalspecimenprocessor.exception.EqualityParsingException;
+import eu.dissco.core.digitalspecimenprocessor.schema.EntityRelationship;
+import java.util.HashSet;
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class EqualityService {
+
+  private final Configuration jsonPathConfig;
+  private final ObjectMapper mapper;
+  private static final List<String> IGNORED_FIELDS = List.of(
+      "dcterms:created",
+      "dcterms:modified",
+      "dwc:relationshipEstablishedDate"
+  );
+
+  public boolean isEqual(DigitalSpecimenWrapper currentDigitalSpecimenWrapper,
+      DigitalSpecimenWrapper digitalSpecimenWrapper,
+      DigitalMediaProcessResult digitalMediaProcessResult) {
+    return specimensAreEqual(currentDigitalSpecimenWrapper, digitalSpecimenWrapper)
+        && digitalMediaProcessResult.newMedia().isEmpty()
+        && digitalMediaProcessResult.tombstoneMedia().isEmpty();
+  }
+
+  private boolean specimensAreEqual(DigitalSpecimenWrapper currentDigitalSpecimenWrapper,
+      DigitalSpecimenWrapper digitalSpecimenWrapper) {
+    if (currentDigitalSpecimenWrapper == null
+        || currentDigitalSpecimenWrapper.attributes() == null) {
+      return false;
+    }
+    try {
+      verifyOriginalData(currentDigitalSpecimenWrapper, digitalSpecimenWrapper);
+      var jsonCurrentSpecimen = normaliseJsonNode(
+          mapper.valueToTree(currentDigitalSpecimenWrapper.attributes()), true);
+      var jsonSpecimen = normaliseJsonNode(mapper.valueToTree(digitalSpecimenWrapper.attributes()),
+          true);
+      var isEqual = jsonCurrentSpecimen.equals(jsonSpecimen);
+      if (!isEqual) {
+        log.debug("Specimen {} has changed. JsonDiff: {}",
+            currentDigitalSpecimenWrapper.physicalSpecimenID(),
+            JsonDiff.asJson(jsonCurrentSpecimen, jsonSpecimen));
+      }
+      return isEqual;
+    } catch (JsonProcessingException e) {
+      log.error("Unable to re-serialize JSON. Can not determine equality.", e);
+      return false;
```

Comment:
this should never happen, it doesn't make sense for a JPE to be thrown by re-serializing these jsons

Reply:
So you think that this response is ok? Or should I throw an IllegalStateException